modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformaceStatisticsProcessor.java

```diff
@@ -0,0 +1,262 @@
+        write(writer -> writer.task(sesId, taskName, startTime, duration, affPartId));
+    }
+
+    /**
+     * @param sesId Session id.
+     * @param queuedTime Time job spent on waiting queue.
+     * @param startTime Start time in milliseconds.
+     * @param duration Job execution time.
+     * @param timedOut {@code True} if job is timed out.
+     */
+    public void job(IgniteUuid sesId, long queuedTime, long startTime, long duration, boolean timedOut) {
+        write(writer -> writer.job(sesId, queuedTime, startTime, duration, timedOut));
+    }
+
+    /**
+     * Starts collecting performance statistics.
+     *
+     * @throws IgniteCheckedException If starting failed.
+     */
+    public void startCollectStatistics() throws IgniteCheckedException {
+        A.notNull(metastorage, "Metastorage not ready. Node not started?");
+
+        if (!allNodesSupports(ctx.discovery().allNodes(), IgniteFeatures.PERFORMANCE_STATISTICS))
+            throw new IllegalStateException("Not all nodes in the cluster support collecting performance statistics.");
+
+        if (ctx.isStopping())
+            throw new NodeStoppingException("Operation has been cancelled (node is stopping)");
+
+        metastorage.write(PERFORMANCE_STAT_ENABLED_PREFIX, true);
+    }
+
+    /**
+     * Stops collecting performance statistics.
+     *
+     * @throws IgniteCheckedException If stopping failed.
+     */
+    public void stopCollectStatistics() throws IgniteCheckedException {
+        A.notNull(metastorage, "Metastorage not ready. Node not started?");
+
+        if (ctx.isStopping())
+            throw new NodeStoppingException("Operation has been cancelled (node is stopping)");
+
+        metastorage.write(PERFORMANCE_STAT_ENABLED_PREFIX, false);
+    }
+
+    /** @return {@code True} if collecting performance statistics is enabled. */
+    public boolean enabled() {
+        return enabled;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onKernalStop(boolean cancel) {
+        if (enabled())
+            stopWriter();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onDisconnected(IgniteFuture<?> reconnectFut) {
+        if (enabled())
+            stopWriter();
+    }
+
+    /** Starts or stops collecting statistics on metastorage update. */
+    private void onMetastorageUpdate(boolean start) {
+        ctx.closure().runLocalSafe(() -> {
+            if (start)
+                startWriter();
+            else
+                stopWriter();
+        });
+    }
+
+    /** Starts performance statistics writer. */
+    private void startWriter() {
+        try {
+            synchronized (this) {
+                if (enabled)
+                    return;
+
+                writer = new FilePerformanceStatisticsWriter(ctx);
+
+                writer.start();
+
+                enabled = true;
+
+                log.info("Performance statistics writer started.");
```

Comment:
We should log outside of the `synchronized` section.

Reply:
Done.